### ISSUE TITLE
`--ciphers` may not properly set `--tls13-ciphers`

### ISSUE THREAD

**Issue by MasterInQuestion** [curl/curl#13873]
The current behavior (["--ciphers"](https://curl.se/docs/manpage.html#--ciphers), "--tls13-ciphers" each independent) may feel tricky.
    Would it be more appropriate to readapt the design?
    (or maybe better explain the rationale..?)

    E.g.
    `--tls13-ciphers "TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256"`
    `--ciphers "TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256"`
    .
    `--tls13-ciphers "TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256" --ciphers "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"`
    `--ciphers "TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"`

    See also:
    "--ciphers" may include other, unspecified ciphers in TLS handshake
    [ https://github.com/curl/curl/issues/7198](https://github.com/curl/curl/issues/7198)
    .
    Clarify setting TLS 1.3 ciphers using different backends
    [ https://github.com/curl/curl/issues/3938](https://github.com/curl/curl/issues/3938)
    [ https://github.com/curl/curl/issues/3077#issuecomment-426561378](https://github.com/curl/curl/issues/3077#issuecomment-426561378)
    [ https://github.com/curl/curl/issues/2435](https://github.com/curl/curl/issues/2435)

    Test site: https://browserleaks.com/tls

**Comment by MasterInQuestion** [comment]
What's the expectation for the "--ciphers" alike..?
    [ https://github.com/curl/curl/issues/1316#issuecomment-285779700](https://github.com/curl/curl/issues/1316#issuecomment-285779700)

    Mere pass-through?
    https://www.openssl.org/docs/manmaster/man1/openssl-ciphers.html#CIPHER-LIST-FORMAT

**Comment by jan2000** [comment]
The situation with the cipher options is not as straightforward as one might expect, because handling of these options differ for each SSL backends that libcurl is build against. For some backends (e.g. mbedtls) the `--cipher` option is parsed by libcurl itself, for others (e.g. openssl, wolfssl) it is passed through directly to the SSL backend.

Due to this the format of the `--cipher` option can differ between backends. While most backends support a list of openssl names[^1] (e.g. `ECDHE-RSA-AES128-GCM-SHA256`), some might also support IANA names (e.g. `TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256`). Some backends expect an all different format all together (e.g. schannel).

When TLSv1.3 was introduced, openssl added a separate option in their api for setting TLSv1.3 ciphers, so libcurl accommodated for that by introducing the `--tls13-ciphers` option. But then when other SSL backends gained TLSv1.3 support they did not follow suit (e.g. wolfssl, mbedtls), and thus their TLSv1.3 ciphers must set by the `--cipher` option next to the TLSv1.0/TLSv1.1/TLSv1.2 ciphers.

This has also the effect that when using the openssl backend, setting the `--cipher` option but omitting the `--tls13-ciphers` option will cause the default TLSv1.3 ciphers to be used. But when using the mbedtls backend only the ciphers in the `--cipher` option will be used.

So the cipher options are not as consistent as one might hope for, due to the idiosyncrasies of the different SSL backends. The situation could be improved, for example libcurl could pre-parse the options so that the expected format and behavior could be made more consistent across the board. That would be not without downsides however (BC breaking changes, extra overhead, added complexity in libcurl), and even then it would not be a panacea for all SSL backends.

For the moment the best you can do, is to check which SSL backend is being used and adjust the cipher options according to the documentation: https://curl.se/docs/ssl-ciphers.html
You might also gain some extra insight by looking at the unit test for the cipher options: `test_17_07_ssl_ciphers` in https://github.com/curl/curl/blob/master/tests/http/test_17_ssl_use.py

[^1]: Notice that for the TLSv1.3 ciphers there are no openssl names. The 1.3 cipher names can be easily told apart from the 1.0/1.1/1.2 IANA names, because all 1.0/1.1/1.2 names contain `WITH_`, while 1.3 names do not.

**Comment by MasterInQuestion** [comment]
So simply put, the current situation is simply havoc...
    For the moment, the most viable fix is better documentation to this situation.

    Making &#x60;curl&#x60; to wrap-around such idiosyncrasies is possible and should perfectly work mostly.
    The primary issue is the complexity of the implementation bother.

**Comment by humbertocastelo** [comment]
`--tls13-ciphers` works only if `--tlsv1.3` is present in the command line, example:

```
--tls13-ciphers 'TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384:ECDHE-ECDSA-AES128-GCM-SHA256' \
--tlsv1.3 \
```

`--ciphers` always works, example:

`--ciphers 'TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384:ECDHE-ECDSA-AES128-GCM-SHA256'`

Server receives:

`TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:TLS_EMPTY_RENEGOTIATION_INFO_SCSV`

This value **`TLS_EMPTY_RENEGOTIATION_INFO_SCSV`** is appended to the end of the cipher list, it seems like it is mandatory and it is impossible to turn off this behavior.

Version:

```
curl 7.81.0 (x86_64-pc-linux-gnu) libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.1.0 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.17
Release-Date: 2022-01-05
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd
```

This has probably already been corrected in newer versions, but I thought it appropriate to leave the comment here, as I found some mentions of this issue.

**Comment by MasterInQuestion** [comment]
"TLS_EMPTY_RENEGOTIATION_INFO_SCSV" is not directly related to this issue. (CLI of &#x60;curl&#x60;)

    I have a writing pending for the "TLS_EMPTY_RENEGOTIATION_INFO_SCSV" trouble.
    Meanwhile you may look at:
    https://archive.org/details/4431785826199-Fix-Discord-Zendesk-Sign-In
    .
    Essential cause: OpenSSL and upstream (TLS standard).

**Comment by MasterInQuestion** [comment]
Also I believe using "--ciphers" to set "--tls13-ciphers" is no-op:
    The default TLS 1.3 Cipher Suites would be then used.

**Comment by icing** [comment]
Looking into the details, I agree that the user experience is confusing and that `--tls13-ciphers` should probably not exist. It exists only for OpenSSL (+ forks) and Schannel. I cannot say much about the latter.

OpenSSL just evaded to define what its cipher string should mean when mixing 1.2 and 1.3 ciphers. It added a separate method, leaving it up to the application to solve the usability. `curl` not wanting to interpret cipher strings by itself, added then another curl option.

Now, this is all understandable, but the user interface just sucks. What can we do? The following restrictions apply:

0. all existing curl command lines **that work now** need to continue working as before. That is our backwards compatibility promise and we need to keep that.
1.  command lines that now do **not** work can be made working with sensible behaviour.
2. curl must not reject any ciphers it does not "know", if it can avoid it. The reason here is that TLS backends may get new ciphers enabled (see quantum thingies) and curl should not prohibit their use.

Given that, I propose the following:

- if both `--ciphers` and `--tls13-ciphers` are specified, the behaviour stays as it is.
- if min TLS version is specified and is < 1.3, the behaviour stays as it is.
- if the max TLS version is specified and is < 1.3, the behaviour stays as it is.

tl;dr: if someone is very specific already, do not interfere.

- If someone specifies they want to do `--tlsv1.3` (or higher) and only uses `--ciphers`, interpret those as if `--tls13-ciphers` were given.
- if someone only uses `--ciphers` without any restrictions on TLS version, try setting the ciphers as 1.2 ciphers. If OpenSSL refuses this, try setting them as 1.3 ciphers. Should that succeed, adjust the minimum TLS version to 1.3.

This would mean a command line like:

```sh
> curl --ciphers TLS_AES_256_GCM_SHA384 https://curl.se
```

would do a TLS v1.3+ handshake with only that cipher and fail a v1.2 handshake. My guess is that is exactly what a user would expect from that command.

**Comment by icing** [comment]
Proposed change in #14106. Please check this, when you find the time, and let us know if this works for you.

**Comment by MasterInQuestion** [comment]
Your current change seems specifically targets:
    [ https://github.com/curl/curl/issues/13873#issuecomment-2183965270](https://github.com/curl/curl/issues/13873#issuecomment-2183965270)
    ("--ciphers" no-op for "--tls13-ciphers")

    Should work.

**Comment by jan2000** [comment]
For curl 8.10.0 improvements have been made to address some issues regarding the cipher options:

With #14384, #14385 the mbedTLS and wolfSSL backends support the `--tls13-ciphers` option. Thus, from curl 8.10.0 onwards all SSL backends (that internally support specifying tls1.3 ciphers) can have their tls1.3 ciphers set with the `--tls13-ciphers` option. With #14406 warnings are now given by curl when the SSL backend does not support the `--tls13-ciphers` and/or `--ciphers` options. And with #14407 and #14460 the documentation of the cipher options has been updated.

As mentioned here before and in #14106 it is not possible to have the `--ciphers` option set both tls1.2 (1.1, 1.0) and tls1.3 ciphers consistently across the supported SSL backends, without mayor revisions and BC breaking changes. So, the proper way to set ciphers is to use the `--cipher` option to set tls1.2 (1.1, 1.0) ciphers and to use `--tls13-ciphers` for the tls1.3 ciphers. See the updated "SSL Ciphers" page for more information: https://curl.se/docs/ssl-ciphers.html

AFAIC there is not much more that can be done for this issue, and thus can be closed.
